flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java

```diff
@@ -662,14 +668,35 @@ private FunctionDefinition getFunctionDefinition(String name, CatalogFunction fu
             // directly.
             return ((InlineCatalogFunction) function).getDefinition();
         }
+        // If the resource of UDF used is not empty, register it to classloader before
+        // validate.
+        registerFunctionResource(name, function.getFunctionResources());
+
         return UserDefinedFunctionHelper.instantiateFunction(
-                classLoader,
+                resourceManager.getUserClassLoader(),
                 // future
                 config,
                 name,
                 function);
     }
 
+    private void registerFunctionResource(String functionName, List<ResourceUri> resourceUris) {
+        if (!resourceUris.isEmpty()) {
+            resourceUris.forEach(
+                    resourceUri -> {
+                        try {
+                            resourceManager.registerResource(resourceUri);
+                        } catch (IOException e) {
+                            throw new ValidationException(
```

Comment:
I think we should do something like transaction to remove all downloaded resources if fails.